File: src/HolzShots.Legacy/Drawing/Tools/Blur.vb
Imports System.Drawing.Drawing2D
Imports HolzShots.Drawing.Tools.UI
Imports HolzShots.UI.Controls

Namespace Drawing.Tools
    Friend Class Blur
        Implements ITool(Of BlurSettings)

        ReadOnly _thePen As New Pen(Color.Red) With {.DashStyle = DashStyle.Dash}
        Private Shared ReadOnly CursorInstance As New Cursor(My.Resources.crossMedium.GetHicon())
        Public ReadOnly Property Cursor As Cursor = CursorInstance Implements ITool(Of BlurSettings).Cursor
        Public ReadOnly Property ToolType As PaintPanel.ShotEditorTool = PaintPanel.ShotEditorTool.Blur Implements ITool(Of BlurSettings).ToolType

        Private ReadOnly _settingsControl As ISettingsControl(Of BlurSettings)
        Public ReadOnly Property SettingsControl As ISettingsControl(Of BlurSettings) Implements ITool(Of BlurSettings).SettingsControl
            Get
                Return _settingsControl
            End Get
        End Property
        Public Property BeginCoordinates As Point Implements ITool(Of BlurSettings).BeginCoordinates
        Public Property EndCoordinates As Point Implements ITool(Of BlurSettings).EndCoordinates

        Sub New()
            _settingsControl = New BlurSettingsControl(BlurSettings.Default)
        End Sub

        Public Sub LoadInitialSettings() Implements ITool(Of BlurSettings).LoadInitialSettings
            If My.Settings.BlurFactor > BlurSettings.MaximumDiameter OrElse My.Settings.BlurFactor <= BlurSettings.MinimumDiameter Then
                My.Settings.BlurFactor = BlurSettings.Default.Diameter
                My.Settings.Save()
            End If

            With _settingsControl.Settings
                .Diameter = My.Settings.BlurFactor
            End With
        End Sub

        Public Sub PersistSettings() Implements ITool(Of BlurSettings).PersistSettings
            With _settingsControl.Settings
                My.Settings.BlurFactor = .Diameter
            End With
        End Sub

        Public Sub RenderFinalImage(ByRef rawImage As Image, sender As PaintPanel) Implements ITool(Of BlurSettings).RenderFinalImage
            If rawImage Is Nothing Then
                Return
            End If

            Dim rawSrcRect As New Rectangle(
                Math.Min(BeginCoordinates.X, EndCoordinates.X),
                Math.Min(BeginCoordinates.Y, EndCoordinates.Y),
                Math.Abs(BeginCoordinates.X - EndCoordinates.X),
                Math.Abs(BeginCoordinates.Y - EndCoordinates.Y)
            )

            If rawSrcRect.X < 0 Then
                rawSrcRect.Width += rawSrcRect.X
                rawSrcRect.X = 0
            End If
            If rawSrcRect.Y < 0 Then
                rawSrcRect.Height += rawSrcRect.Y
                rawSrcRect.Y = 0
            End If

            If rawSrcRect.Width = 0 OrElse rawSrcRect.Height = 0 Then Exit Sub

            Using g As Graphics = Graphics.FromImage(rawImage)
                Dim settings = _settingsControl.Settings

                Using img As Bitmap = BlurImage(rawImage, settings.Diameter, rawSrcRect)
                    If img Is Nothing Then Exit Sub

                    g.FillRectangle(Brushes.White, rawSrcRect)
                    g.CompositingMode = CompositingMode.SourceOver
                    g.DrawImage(img, rawSrcRect)
                End Using
            End Using
        End Sub

        Public Sub RenderPreview(rawImage As Image, g As Graphics) Implements ITool(Of BlurSettings).RenderPreview
            If rawImage Is Nothing Then
                Return
            End If

            Dim rawSrcRectangle As New Rectangle(
                Math.Min(BeginCoordinates.X, EndCoordinates.X),
                Math.Min(BeginCoordinates.Y, EndCoordinates.Y),
                Math.Abs(BeginCoordinates.X - EndCoordinates.X),
                Math.Abs(BeginCoordinates.Y - EndCoordinates.Y)
            )

            If rawSrcRectangle.X < 0 Then
                rawSrcRectangle.Width += rawSrcRectangle.X
                rawSrcRectangle.X = 0
            End If

            If rawSrcRectangle.Y < 0 Then
                rawSrcRectangle.Height += rawSrcRectangle.Y
                rawSrcRectangle.Y = 0
            End If

            If rawSrcRectangle.Width = 0 OrElse rawSrcRectangle.Height = 0 Then Exit Sub

            Dim settings = _settingsControl.Settings

            Using img As Bitmap = BlurImage(rawImage, settings.Diameter, rawSrcRectangle)
                If img Is Nothing Then Exit Sub

                g.CompositingMode = CompositingMode.SourceOver
                g.FillRectangle(Brushes.White, rawSrcRectangle)
                g.DrawImage(img, rawSrcRectangle)
            End Using
            g.DrawRectangle(_thePen, rawSrcRectangle)
        End Sub

        Private Shared Function BlurImage(img As Image, factor As Integer, rawSrcRect As Rectangle) As Bitmap
            Dim width As Integer = CInt(Math.Round(Math.Ceiling(rawSrcRect.Width / factor)))
            Dim height As Integer = CInt(Math.Round(Math.Ceiling(rawSrcRect.Height / factor)))

            If width <= 0 OrElse height <= 0 Then Return Nothing

            Using smallBitmap As New Bitmap(width, height)
                Using g As Graphics = Graphics.FromImage(smallBitmap)
                    g.CompositingMode = CompositingMode.SourceOver
                    g.InterpolationMode = InterpolationMode.HighQualityBilinear
                    g.DrawImage(
                        img,
                        New Rectangle(0, 0, smallBitmap.Width, smallBitmap.Height),
                        rawSrcRect,
                        GraphicsUnit.Pixel
                    )
                End Using

                Using secondBitmap As New Bitmap(rawSrcRect.Width, rawSrcRect.Height)
                    Using g As Graphics = Graphics.FromImage(secondBitmap)
                        g.CompositingMode = CompositingMode.SourceOver
                        g.InterpolationMode = InterpolationMode.HighQualityBilinear
                        g.DrawImage(
                            smallBitmap,
                            New Rectangle(0, 0, rawSrcRect.Width + factor, rawSrcRect.Height + factor),
                            New Rectangle(0, 0, smallBitmap.Width, smallBitmap.Height),
                            GraphicsUnit.Pixel
                        )
                    End Using
                    Return DirectCast(secondBitmap.Clone(), Bitmap)
                End Using
            End Using
        End Function
        Public Sub MouseOnlyMoved(rawImage As Image, ByRef currentCursor As Cursor, e As MouseEventArgs) Implements ITool(Of BlurSettings).MouseOnlyMoved
            ' Nothing to do here
        End Sub
        Public Sub MouseClicked(rawImage As Image, e As Point, ByRef currentCursor As Cursor, trigger As Control) Implements ITool(Of BlurSettings).MouseClicked
            ' Nothing to do here
        End Sub
        Public Sub Dispose() Implements ITool(Of BlurSettings).Dispose
        End Sub
    End Class
End Namespace

File: src/HolzShots.LegacyNew/Drawing/Tools/Blur.cs
using System.Drawing.Drawing2D;
using System.Numerics;
using HolzShots.Drawing.Tools.UI;

namespace HolzShots.Drawing.Tools;

public class Blur : ITool<BlurSettings>
{
    private readonly Pen _thePen = new Pen(Color.Red) { DashStyle = DashStyle.Dash };
    private static readonly Cursor CursorInstance = new Cursor(Properties.Resources.crossMedium.GetHicon());
    public Cursor Cursor { get; } = CursorInstance;
    public ShotEditorTool ToolType { get; } = ShotEditorTool.Blur;

    public ISettingsControl<BlurSettings> SettingsControl { get; } = new BlurSettingsControl(BlurSettings.Default);
    public Vector2 BeginCoordinates { get; set; }
    public Vector2 EndCoordinates { get; set; }

    public void LoadInitialSettings()
    {
        if (Properties.Settings.Default.BlurFactor > BlurSettings.MaximumDiameter || Properties.Settings.Default.BlurFactor <= BlurSettings.MinimumDiameter)
        {
            Properties.Settings.Default.BlurFactor = BlurSettings.Default.Diameter;
            Properties.Settings.Default.Save();
        }

        var settings = SettingsControl.Settings;
        settings.Diameter = Properties.Settings.Default.BlurFactor;
    }

    public void PersistSettings()
    {
        var settings = SettingsControl.Settings;
        Properties.Settings.Default.BlurFactor = settings.Diameter;
    }

    public void RenderFinalImage(ref Image rawImage)
    {
        if (rawImage == null)
            return;

        var rawSrcRect = Rectangle.Round(new RectangleF(
            Math.Min(BeginCoordinates.X, EndCoordinates.X),
            Math.Min(BeginCoordinates.Y, EndCoordinates.Y),
            Math.Abs(BeginCoordinates.X - EndCoordinates.X),
            Math.Abs(BeginCoordinates.Y - EndCoordinates.Y)
        ));

        if (rawSrcRect.X < 0)
        {
            rawSrcRect.Width += rawSrcRect.X;
            rawSrcRect.X = 0;
        }

        if (rawSrcRect.Y < 0)
        {
            rawSrcRect.Height += rawSrcRect.Y;
            rawSrcRect.Y = 0;
        }

        if (rawSrcRect.Width == 0 || rawSrcRect.Height == 0)
            return;

        using var g = Graphics.FromImage(rawImage);
        var settings = SettingsControl.Settings;

        using var img = BlurImage(rawImage, settings.Diameter, rawSrcRect);
        if (img == null)
            return;

        g.FillRectangle(Brushes.White, rawSrcRect);
        g.CompositingMode = CompositingMode.SourceOver;
        g.DrawImage(img, rawSrcRect);
    }

    public void RenderPreview(Image rawImage, Graphics g)
    {
        if (rawImage == null)
            return;

        var rawSrcRectangle = Rectangle.Round(new RectangleF(
            Math.Min(BeginCoordinates.X, EndCoordinates.X),
            Math.Min(BeginCoordinates.Y, EndCoordinates.Y),
            Math.Abs(BeginCoordinates.X - EndCoordinates.X),
            Math.Abs(BeginCoordinates.Y - EndCoordinates.Y)
        ));

        if (rawSrcRectangle.X < 0)
        {
            rawSrcRectangle.Width += rawSrcRectangle.X;
            rawSrcRectangle.X = 0;
        }

        if (rawSrcRectangle.Y < 0)
        {
            rawSrcRectangle.Height += rawSrcRectangle.Y;
            rawSrcRectangle.Y = 0;
        }

        if (rawSrcRectangle.Width == 0 || rawSrcRectangle.Height == 0)
            return;

        var settings = SettingsControl.Settings;

        using var img = BlurImage(rawImage, settings.Diameter, rawSrcRectangle);
        if (img == null)
            return;

        g.CompositingMode = CompositingMode.SourceOver;
        g.FillRectangle(Brushes.White, rawSrcRectangle);
        g.DrawImage(img, rawSrcRectangle);
        g.DrawRectangle(_thePen, rawSrcRectangle);
    }

    private static Bitmap BlurImage(Image img, int factor, Rectangle rawSrcRect)
    {
        var width = Convert.ToInt32(Math.Round(Math.Ceiling(rawSrcRect.Width / (double)factor)));
        var height = Convert.ToInt32(Math.Round(Math.Ceiling(rawSrcRect.Height / (double)factor)));

        if (width <= 0 || height <= 0)
            return null; // TODO Change to default(_) if this is not a reference type

        using var smallBitmap = new Bitmap(width, height);
        using var g0 = Graphics.FromImage(smallBitmap);

        g0.CompositingMode = CompositingMode.SourceOver;
        g0.InterpolationMode = InterpolationMode.HighQualityBilinear;
        g0.DrawImage(img, new Rectangle(0, 0, smallBitmap.Width, smallBitmap.Height), rawSrcRect, GraphicsUnit.Pixel);

        using var secondBitmap = new Bitmap(rawSrcRect.Width, rawSrcRect.Height);
        using var g1 = Graphics.FromImage(secondBitmap);

        g1.CompositingMode = CompositingMode.SourceOver;
        g1.InterpolationMode = InterpolationMode.HighQualityBilinear;
        g1.DrawImage(
            smallBitmap,
            new Rectangle(0, 0, rawSrcRect.Width + factor, rawSrcRect.Height + factor),
            new Rectangle(0, 0, smallBitmap.Width, smallBitmap.Height),
            GraphicsUnit.Pixel
        );
        return (Bitmap)secondBitmap.Clone();
    }

    public void MouseOnlyMoved(Image rawImage, ref Cursor currentCursor, MouseEventArgs e) { }
    public void MouseClicked(Image rawImage, Vector2 e, ref Cursor currentCursor, Control trigger) { }
    public void Dispose() { }
}
